### PR TITLE
feature #434: add auth with cookies

### LINKED ISSUE #434: [ENHANCEMENT] Rewrite Auth to Cookies
### **Describe the solution you'd like**
We need to enhance the current authentication logic to improve security, performance, and maintainability.

## Frontend Changes
1. **Token Handling**:
   - Transition from using `localStorage` for storing JWT tokens to using **cookies** with `HttpOnly` attribute (if possible). This will reduce the risk of XSS attacks and improve the security of session management.
   - Update all API calls and utilities to retrieve tokens from cookies rather than `localStorage`.

2. **Axios Interceptors for Token Refreshing**:
   - Implement **Axios interceptors** to automatically refresh tokens when the access token expires, streamlining the token management process and eliminating the need for manual token handling in each API call.

3. **Authentication Context**:
   - Refactor the `AuthenticationProvider` to reduce unnecessary state updates and re-renders, improving performance.
   - Introduce better error handling, particularly around token expiration and refreshing (with axios interceptors)

4. **API Error Handling**:
   - Centralize error handling across the app by ensuring all API calls that require an authorization header dynamically retrieve tokens from cookies, eliminating the reliance on `localStorage`.

5. **React Query Optimization**:
   - Optimize the usage of `react-query` by leveraging conditional fetching. Ensure queries are only executed when JWT tokens are available, thus reducing unnecessary API requests and improving performance.

6 **Remove Blocking Rendering**:
   - Currently we do NOT show anything while getMe method is being called because of the conditional rendering in AuthContext, as result user cannot see anything before user is fetched. Your task is to remove conditional rendering with children 

By implementing these improvements, we aim to create a more secure, scalable, performant, and maintainable authentication flow while continuing to use our existing libraries and frameworks.

### 1. **Token Handling in Cookies**
   - **Objective**: Transition from using the `Authorization` header with a Bearer token to using cookies for authentication.
   - **Implementation**:
     - Update the backend (NestJS) to parse `access_token` and `refresh_token` from cookies instead of the `Authorization` header.
     - Ensure the cookies are secure, with the `HttpOnly` attribute set to protect against XSS attacks.
     - Modify the authentication middleware/guards to validate tokens extracted from cookies.

### 2. **Token Refresh Logic**
   - **Objective**: Support the frontend's Axios interceptor mechanism for token refreshing.
   - **Implementation**:
     - Implement backend logic to handle token refresh requests by validating the `refresh_token` from the cookie and issuing a new `access_token`.
     - Ensure the new tokens are set in the body.

---

### **Additional context**
- This improvement enhances the security of our web application by reducing the vulnerability to XSS attacks.
- This refactoring will improve performance by minimizing unnecessary re-renders, reducing redundant API calls, and ensuring a more streamlined token management process using Axios interceptors.

### PR DESCRIPTION
**Pull request by sifes**
front:
- update useAuthentication and authenticationContext
- remove blocking rendering
- remove getAuthHeader
- add WithCredentials: true in axios
- add interceptor in axios for updating access_token
- remove AuthService

back:
- replaced bearer auth with cookie auth
- created refresh strategy and guard

closes #434

### CODE REVIEW

**Review by dmykyr — APPROVED** [review]

File: fictadvisor-web/src/app/(auth)/register/components/telegram-register-buttons/TelegramRegisterButtons.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  const handleClick = async () => {
+    await TelegramService.redirectToRegisterBot();
+  };
+
+  return hasTelegram ? (
+    <Typography variant="h6" sx={stylesMUI.telegramConnected}>
+      Telegram приєднано, дозаповни усі поля
+    </Typography>
+  ) : (
```
Comment (Markusplay) [inline fictadvisor-web/src/app/(auth)/register/components/telegram-register-buttons/TelegramRegisterButtons.tsx:27]: Move to if statement

File: fictadvisor-web/src/components/common/layout/header/utils/transformData.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ const roleMapper: Record<string, string> = {
   [GroupRoles.STUDENT]: 'Студент',
 };
 
-const transformData = (user?: User): TransformedUser => {
+export const transformUserData = (user: User): TransformedUser => {
   const name = [user?.lastName, user?.firstName, user?.middleName].join(' ');
```
Comment (Markusplay) [inline fictadvisor-web/src/components/common/layout/header/utils/transformData.ts:14]: do we need to check if user is defined, if we already pass non null object?

File: fictadvisor-web/src/lib/api/instance.ts [hunk 3]
```diff
@@ -1,6 +1,43 @@
...
+    return response;
+  },
+  async error => {
+    if (error.response?.status === 401) {
+      try {
+        await refreshToken();
+        const originalRequest = error.config;
+        return await axios(originalRequest);
+      } catch (error_1) {
+        throw error_1;
```
Comment (Markusplay) [inline fictadvisor-web/src/lib/api/instance.ts:24]: do not throw error in catch

File: fictadvisor-web/src/lib/providers/authentication/AuthenticationProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+'use client';
+import { createContext, FC, PropsWithChildren, useContext } from 'react';
+import { useQuery } from '@tanstack/react-query';
+
+import AuthAPI from '@/lib/api/auth/AuthAPI';
+import { isServer } from '@/lib/constants/isServer';
+import { getClientCookie } from '@/lib/utils/getClientCookie';
+import { AuthenticationContext, AuthTokenType } from '@/types/auth';
+
+const authenticationContext = createContext<AuthenticationContext>({
```
Comment (Markusplay) [inline fictadvisor-web/src/lib/providers/authentication/AuthenticationProvider.tsx:10]: wtf is it

**Review by Markusplay — CHANGES_REQUESTED** [review]

File: fictadvisor-web/src/lib/providers/authentication/AuthenticationProvider.tsx (hunk 4 above)
Comment (sifes) [reply fictadvisor-web/src/lib/providers/authentication/AuthenticationProvider.tsx:10]: wait, actually, whats wrong with this

**Review by Markusplay — APPROVED** [review]

**Pull request merged**